### DISCUSSION TITLE
Most efficient way to process annotations

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-jdt/eclipse.jdt.core#523] [General]
Currently PDE uses a [OSGiAnnotationsCompilationParticipant](https://github.com/eclipse-pde/eclipse.pde/blob/6249405a278ccef2b2f02fe21869c9517b121208/ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/annotations/OSGiAnnotationsCompilationParticipant.java#L39)(the same technique is used in [DSAnnotationCompilationParticipant](https://github.com/eclipse-pde/eclipse.pde/blob/master/ds/org.eclipse.pde.ds.annotations/src/org/eclipse/pde/ds/internal/annotations/DSAnnotationCompilationParticipant.java))  to find out if an annotation is present in the compiled class and then analyze it to possibly generate some code.

This has add a noticeable performance overhead most likely because of the AST creation, and even if we could optimize the current processing probably here (it is currently rather straight forward than optimized for performance), I wonder if there is maybe a better way.

What bugs me a bit here is that obviously JDT already knows that there ARE annotations present, but it does not pass the data in a way I can access e.g. *what* annotations are present (what would make it much more easier to decide if there is actually something to do, but probably this is because it is actually easy and performant to gather this data anyway?

**Comment by HannesWell** [comment]
@iloveeclipse maybe you can tell more about this or ping somebody who knows more?
